cli/src/main/java/com/box/l10n/mojito/cli/filefinder/DirectoryScanUtils.java

```diff
@@ -0,0 +1,124 @@
+   * the directoriesExcludePatterns list is non-null, the directory must not match any of the
+   * "exlucde" patterns. If the directoriesIncludePatterns list is non-null, the directory will have
+   * to match at least one of the "include" patterns.
+   */
+  public boolean shouldScan(Path dir) {
+    return rootDirectory.equals(dir) || (!isExcludedDirectory(dir) && isIncludedDirectory(dir));
+  }
+
+  private boolean isExcludedDirectory(Path dir) {
+    boolean exclude = false;
+    if (directoriesExcludePatterns != null) {
+      String[] relativePathSplit = getRelativePathSplit(dir);
+      exclude =
+          directoriesExcludePatterns.stream()
+              .anyMatch(
+                  s ->
+                      s.length == relativePathSplit.length
+                          && DirectoryScanUtils.shouldScanDirectory(s, relativePathSplit));
+    }
+    return exclude;
+  }
+
+  private boolean isIncludedDirectory(Path dir) {
+    boolean include = true;
+    if (directoriesIncludePatterns != null) {
+      String[] relativePathSplit = getRelativePathSplit(dir);
+      include =
+          directoriesIncludePatterns.stream()
+              .anyMatch(s -> DirectoryScanUtils.shouldScanDirectory(s, relativePathSplit));
+    }
+    return include;
+  }
+
+  private String[] getRelativePathSplit(Path dir) {
+    Path relativePath = rootDirectory.relativize(dir);
+    String[] relativePathSplit = relativePath.toString().split("/");
+    return relativePathSplit;
+  }
+
+  /**
+   * Indicates if a directory should be scanned given a pattern. A directory should be scanned if
+   * its sub paths matches the ones in the directory pattern. If the directory is shorter than the
+   * pattern, the directory just needs to match the beginning of the pattern.
+   *
+   * <p>The directory pattern is a list of sub paths. A sub path can either be a plain name or a
+   * wildcard "*". The wildcard maps to exactly one sub path. Examples: [ "webapp", "locale" ], [
+   * "*", "locale" ]
+   *
+   * <p>shouldScanDirectory([ "webapp", "locale" ], [ "webapp", "locale" ]) --> true
+   * shouldScanDirectory([ "webapp", "locale" ], [ "webapp", "locale2" ]) --> false
+   * shouldScanDirectory([ "webapp", "locale" ], [ "webapp" ]) --> true, we need to recurse to find
+   * directories that potentially match shouldScanDirectory([ "*", "locale" ], [ "webapp" ]) -->
+   * true shouldScanDirectory([ "*", "locale" ], [ "webapp", "locale" ]) --> true
+   *
+   * <p>Edge cases: - if the directory is an empty array, which is equivalent to ask should we scan
+   * the root, then this will return true since we need to scan the root regardless of the pattern
+   * to get started - if an empty array is provided as pattern it is equivalent to matching the root
+   * and it will always be true regardless of the directory
+   */
+  static boolean shouldScanDirectory(String[] directoryPattern, String[] aDirectory) {
```

Comment:
Should the pattern and directory be passed as one object that has a Pattern and Directory value in a list of similar objects?

I was confused initially that any of the patterns could apply to any of the directories passed in.

Reply:
Also - this method name should be pluralized.

Reply:
Not 100% sure to get your point, more context... This is working on a single directory with a single pattern. Args are string arrays because the pattern and path have been split already. I didn't want to keep splitting the pattern every single time (see splitPatterns() called in the ctor). Does it make sense or you mean something different?